### PR TITLE
Adding class weights to accuracy

### PR DESCRIPTION
**Pull request by sachinruk**
For imbalanced classes it is important to look at accuracy after reweighting each instance. This achieves that. Note that it is not sample weights.

Apologies for not completing the classes with accuracy. Wanted to do only a few changes before I do something fullblown.

## Notes
- I'm not sure what `"multi-dim multi-class"` is, so I have not included weights there.
- I do `weights = class_weights.to(target.device)[target]` incase this is happening in different GPUs. This way we know that weights and targets are always on the same device no matter what.

### CODE REVIEW

File: pytorch_lightning/metrics/functional/accuracy.py [hunk 1]
```diff
@@ -18,23 +18,35 @@
...
-        total = torch.tensor(target.numel(), device=target.device)
+        correct = weights * (preds == target).sum()
+        # total = torch.tensor(target.numel(), device=target.device)
+        total = weights.sum()
     elif mode == "multi-class" or (mode == "multi-dim multi-class" and not subset_accuracy):
-        correct = (preds * target).sum()
-        total = target.sum()
+        correct = (weights * preds * target).sum()
+        total = (weights * target).sum()
     elif mode == "multi-dim multi-class" and subset_accuracy:
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/functional/accuracy.py:50]: maybe @tadejsv can help what the correct formula would be in the last case

File: pytorch_lightning/metrics/functional/accuracy.py [hunk 2]
```diff
@@ -51,6 +63,7 @@ def accuracy(
     preds: torch.Tensor,
     target: torch.Tensor,
     threshold: float = 0.5,
+    class_weights: Optional[torch.Tensor] = None,
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/functional/accuracy.py:66]: remember to add docstring :]

**Review by SkafteNicki — COMMENTED** [review]
Nice addtion. Please also add the changes to the `Accuracy` class and also add a couple of test for this

File: pytorch_lightning/metrics/functional/accuracy.py [hunk 3]
```diff
@@ -18,23 +18,35 @@
...
     if mode == "binary" or (mode == "multi-label" and subset_accuracy):
-        correct = (preds == target).all(dim=1).sum()
-        total = torch.tensor(target.shape[0], device=target.device)
+        correct = (weights * (preds == target).all(dim=1)).sum()
+        # total = torch.tensor(target.shape[0], device=target.device)
+        total = weights.sum()
     elif mode == "multi-label" and not subset_accuracy:
-        correct = (preds == target).sum()
-        total = torch.tensor(target.numel(), device=target.device)
+        correct = weights * (preds == target).sum()
```
Comment (tadejsv) [inline pytorch_lightning/metrics/functional/accuracy.py:44]: this will not work...

File: pytorch_lightning/metrics/functional/accuracy.py [hunk 4]
```diff
@@ -18,23 +18,35 @@
...
+    else:
+        weights = torch.ones_like(target)
+
     if mode == "multi-label" and top_k:
         raise ValueError("You can not use the `top_k` parameter to calculate accuracy for multi-label inputs.")
 
     if mode == "binary" or (mode == "multi-label" and subset_accuracy):
-        correct = (preds == target).all(dim=1).sum()
-        total = torch.tensor(target.shape[0], device=target.device)
+        correct = (weights * (preds == target).all(dim=1)).sum()
```
Comment (tadejsv) [inline pytorch_lightning/metrics/functional/accuracy.py:40]: Weighting should not appear in this case. Subset accuracy means that all elements in each sample must be correctly predicted - there can be no weighting here. And in binary case (1 class) it doesn't make sense anyway

File: pytorch_lightning/metrics/functional/accuracy.py [hunk 5]
```diff
@@ -18,23 +18,35 @@
...
+        # total = torch.tensor(target.shape[0], device=target.device)
+        total = weights.sum()
```
Comment (tadejsv) [inline pytorch_lightning/metrics/functional/accuracy.py:42]: See above

File: pytorch_lightning/metrics/functional/accuracy.py [hunk 6]
```diff
@@ -18,23 +18,35 @@
...
     elif mode == "multi-label" and not subset_accuracy:
-        correct = (preds == target).sum()
-        total = torch.tensor(target.numel(), device=target.device)
+        correct = weights * (preds == target).sum()
+        # total = torch.tensor(target.numel(), device=target.device)
+        total = weights.sum()
     elif mode == "multi-class" or (mode == "multi-dim multi-class" and not subset_accuracy):
-        correct = (preds * target).sum()
-        total = target.sum()
+        correct = (weights * preds * target).sum()
```
Comment (tadejsv) [inline pytorch_lightning/metrics/functional/accuracy.py:48]: This is just weighted recall

**Review by tadejsv — COMMENTED** [review]
I think that adding weighting to this class is not a good idea. Currently the code only makes some sense for the MC case, where if I am not mistaken, it does exactly the same thing as the weighted Recall. My Precision/Recall PR (should be ready today) will add a `average='weighted'` option for that, which should take care of unbalanced datasets. It would be better to consider adding custom weighting there. Besides duplication, this is problematic as there is no agreed upon definition of what a "class accuracy" is in MC case.

And for the ML (without subset accuracy), just weighted Recall would not be an option, as there we also care about correctly predicted negatives. So a generalization there would be `tp + tn / (tp + fp + tn + fn)` (more on that below)

This I think just creates more confusion for the users. I would recommend the following:
- adding custom weighting to Precision/Recall.
- creating a BinaryAccuracy (or something similar) that would be based on StatScores, and explicitly calculated `tp + tn / (tp + fp + tn + fn)` - this would generalize accuracy in ML case. We could then use the standard weighting methods (micro/macro/weighted/samples/none) for it.

Also relevant is the [balanced accuracy](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.balanced_accuracy_score.html#sklearn.metrics.balanced_accuracy_score) from sklearn.

File: pytorch_lightning/metrics/functional/accuracy.py (hunk 3 above)
Comment (sachinruk) [reply pytorch_lightning/metrics/functional/accuracy.py:44]: Why? is it because weights, preds and targets are N x C matrices? I set total to weights.sum() below so takes care of the matrix nature of all 3 tensors.

**Comment by sachinruk** [comment]
closing this. Seems like this won't get approved. All good.
